pkg/driver/controller.go

```diff
@@ -459,10 +456,10 @@ func (d *controllerService) ControllerUnpublishVolume(ctx context.Context, req *
 	volumeID := req.GetVolumeId()
 	nodeID := req.GetNodeId()
 
-	if !d.inFlight.Insert(volumeID) {
+	if !d.inFlight.Insert((volumeID + nodeID)) {
```

Comment:
Update `pkg/driver/internal/inFlight` so that its interface ensures that the right thing is happening, without relying on the best intentions of the caller. 

First, you'll have to decide what the right thing is. It looks like today, a volume can only have one pending operation, regardless of what the operation is. For example, CreateSnapshot will throw an error if there is a Publish or Unpublish in progress. I'm not saying that that's the right behavior. I'm saying that you are quietly changing that behavior. Was that your intention?

Reply:
> First, you'll have to decide what the right thing is.

The CSI specification specifies that a specific operation should be idempotent. CreateSnapshot throwing an error if there is a Publish in progress as you point or more generally speaking unnecessarily locking up the volume is not very efficient or desirable. 

> Was that your intention?

Yes, up until now we've largely been able to get away with using the volumeID as a unique identifier for the inflight map. That is no longer the case because -- as an example -- attempting to attach the same volume to multiple instances at the same time is a valid operation.

> Update pkg/driver/internal/inFlight so that its interface ensures that the right thing is happening, without relying on the best intentions of the caller.

Added a new optional parameter, `extra`, that can be used by callers to ensure uniqueness of the key passed in to `Insert`. Initially thought about enforcing `nodeId` as the string to append to the key but that seems rather restrictive and not appropriate in every case given that this in-flight map is used for all operations in the driver. Ex for snapshot requests, the snapshotID is a valid unique key and the use of extra is not necessary.

Reply:
As I said above, this is a quiet behavior change. But it's a change from one non-compliant behavior to another non-compliant behavior. So I guess it would be out-of-scope for me to complain about that in this PR. But we really should revisit this. The way that `inFlight` is used in this module is neither necessary nor sufficient to achieve idempotency. And the more that I read [the spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), the less I understand what it means by "idempotent". It seems ill-defined--or at least under-specified--for several operations.